records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -0,0 +1,83 @@
+
+## Why this feature is needed
+The purpose of this feature is to keep the bounce rate as low as possible which will mean a positive sending reputation for the GC Notify platform. This will ensure emails are consistently trusted and delivered by email providers.  This can be done by monitoring each service's bounce rate and automatically suspending their service when they reach a certain threshold.  It will also be necessary to provide users with guides to explain these concepts and help to them reduce their bounce rate.
+
+## Proposed solution
+
+### Calculate service bounce rates
+Record the bounce status of every notification and calculate number of emails sent and the bounce rate for each service over the past [bounce-rate-window]
+
+### Notify a service when it is within [bounce-rate-warning-percentage] of the [bounce-rate-threshold]
+When a service is approaching the [bounce-rate-threshold], notify the service owner so they can take action to reduce their bounce rate.  The message should contain a link to the service's dashboard where they can see their bounce rate and take action to reduce it.
+
+#### Minimum volume threshold
+Low-volume senders won't have a large impact on the bounce rate, and notifying them directly may lead to many support calls. In order to limit this, only notify services that are sending more than [bounce-rate-service-volume-minimum] emails in the [bounce-rate-window].
+
+### Suspend a service when it reaches the [bounce-rate-threshold]
+When a service reaches the [bounce-rate-threshold], suspend their service to prevent them from sending any more emails. Notify the service owner that their service has been suspended and provide them with a link to the service's dashboard where they can see their bounce rate and take action to reduce it.
+
+:question: _How to decide when it is appropriate to unsuspend a service?_
+
+### Add bounce rate information to the service dashboad
+Add the bounce rate to the service dashboard so that service owners can see their bounce rate and proactively read the guidance on it so they can prevent their bounce rate from rising in the first place.
+
+### Provide bounce rate guidance
+Provide guidance to service owners to explain bounce rate concepts and how they apply to their service.  Provide guidance on how services can keep their bounce rate low. 
+
+### bounce-rate-threshold calculation options
+When calculating a service's bounce rate, we can either consider them in isolation or in aggregate with other services.  The bounce rate is calculated as a percentage of the total number of emails sent.  If we consider a service in isolation, then the bounce rate will be calculated as a percentage of the total number of emails sent by that service.  If we consider a service in aggregate, then the bounce rate will be calculated as a percentage of the total number of emails sent by all services.
```

Comment:
We do want a bounce rate isolated to a service, this is our need. But we could actually calculate service isolated and global bounce rate, the latter would be useful for admins.

Reply:
Yeah I was trying to balance understandability with usefulness here and also thinking about bounce rates in tow categories: 1) those that affect our overall rate and 2) those that that don't.  

I think the best bet will be to base the dashboard display on the bounce rate in relation to the service's total send rate as this will make the most sense to a user.  In terms of notifying/suspending services however, we could do that for services with high individual bounce rates and ALSO are actually impacting the overall bounce rate.  This would reduce support requests on our end with the trade-off of letting lower-volume services keep high bounce rates while protecting the overall Notify platform by focusing on the most problematic services.

I think the key to this will be devising the `bounce-rate-service-volume-minimum` and `bounce-rate-window` values to ensure we focus on the right services.  I will run a some calculations and add some tables to the ADR to try to illustrate this.

Reply:
Oh I just read that reply. This might be related to another comment I just left and I think it could be problematic and complex to implement. We can discuss in person to have a higher communication bandwidth.

Reply:
Yes let's discuss.  I ran some numbers comparing the idea of individual vs overall bounce rates and you are right it is overly complex for no real added benefit.  I'm going to re-work the ADR to remove this more complex idea.